client/src/controllers/SyncController.test.js

```diff
@@ -0,0 +1,54 @@
+import { Application } from '@hotwired/stimulus';
+import { SlugController } from './SlugController';
+import { SyncController } from './SyncController';
+
+describe('SyncController', () => {
+  let application;
+
+  require('../../../wagtail/admin/static_src/wagtailadmin/js/vendor/urlify')
+    .default;
+
+  beforeEach(() => {
+    application?.stop();
+
+    document.body.innerHTML = `
+        <div>
+            <input
+                type="text"
+                name="title"
+                placeholder="Page title*"
+                maxlength="255"
+                id="id_title"
+                data-controller="w-sync"
+                data-action="focus->w-sync#check change->w-sync#apply cut->w-sync#clear"
+                data-w-sync-delay-value="500"
+                data-w-sync-target-value="body:not(.page-is-live) #id_slug"
+            />
+            <input 
+                id="id_slug" 
+                name="slug" 
+                data-controller="w-slug" 
+                data-action="blur->w-slug#slugify w-sync:check->w-slug#compare w-sync:apply->w-slug#urlify:prevent"
+                data-w-slug-mode-param="urlify"" 
+                type="text" 
+            />
+        </div>`;
+
+    Application.start().register('w-slug', SlugController);
+    Application.start().register('w-sync', SyncController);
+  });
+
+  it('should check if the title field value is the same with the slug value', () => {
+    const slugInput = document.querySelector('#id_slug');
+    const titleInput = document.querySelector('#id_title');
+
+    slugInput.value = 'title testingand slug testing';
+    titleInput.value = 'title testingand slug testing';
+
+    slugInput.dispatchEvent(new CustomEvent('focus'));
+    slugInput.dispatchEvent(new CustomEvent('blur'));
+
+    expect(slugInput.value).toEqual('title-testingand-slug-testing');
+    expect(titleInput.value).toEqual('title-testingand-slug-testing');
+  });
```

Comment:
Here is a starting point test suite with some recommended next steps, key call out here is that we should not have to mock the urlify stuff as the sync controller does not know about urlify, it just knows about dispatching events and sometimes updating some other field.

I hope this is helpful, I know that this is a more complex one so wanted to ensure we cover a range of tests, happy to help further if you get stuck.


```ts
import { Application } from '@hotwired/stimulus';
import { SyncController } from './SyncController';

jest.useFakeTimers();

describe('SyncController', () => {
  let application;

  describe('basic sync between two fields', () => {
    beforeAll(() => {
      application?.stop();

      document.body.innerHTML = `
    <section>
      <input type="text" name="title" id="title" />
      <input
        type="date"
        id="event-date"
        name="event-date"
        value="2025-07-22"
        data-controller="w-sync"
        data-action="change->w-sync#apply cut->w-sync#clear custom:event->w-sync#ping"
        data-w-sync-target-value="#title"
      />
    </section>`;

      application = Application.start();
    });

    it('should dispatch a start event on targeted element', () => {
      const startListener = jest.fn();
      document
        .getElementById('title')
        .addEventListener('w-sync:start', startListener);

      expect(startListener).not.toHaveBeenCalled();

      application.register('w-sync', SyncController);

      expect(startListener).toHaveBeenCalledTimes(1);

      expect(startListener.mock.calls[0][0].detail).toEqual({
        element: document.getElementById('event-date'),
        value: '2025-07-22',
      });
    });

    it('should allow the sync field to apply its value to the target element', () => {
      const changeListener = jest.fn();
      document
        .getElementById('title')
        .addEventListener('change', changeListener);

      expect(document.getElementById('title').value).toEqual('');
      expect(changeListener).not.toHaveBeenCalled();

      const dateInput = document.getElementById('event-date');

      dateInput.value = '2025-05-05';
      dateInput.dispatchEvent(new Event('change'));

      jest.runAllTimers();

      expect(document.getElementById('title').value).toEqual('2025-05-05');
      expect(changeListener).toHaveBeenCalledTimes(1);
    });

    it('should allow for a simple ping against the target field that bubbles', () => {
      const pingListener = jest.fn();
      document.addEventListener('w-sync:ping', pingListener);

      expect(pingListener).not.toHaveBeenCalled();

      const dateInput = document.getElementById('event-date');

      dateInput.dispatchEvent(new CustomEvent('custom:event'));

      expect(pingListener).toHaveBeenCalledTimes(1);
      const event = pingListener.mock.calls[0][0];

      expect(event.target).toEqual(document.getElementById('title'));

      expect(event.detail).toEqual({
        element: document.getElementById('event-date'),
        value: '2025-05-05',
      });
    });

    it('should allow the sync field to clear the value of the target element', () => {
      const changeListener = jest.fn();
      document
        .getElementById('title')
        .addEventListener('change', changeListener);

      expect(document.getElementById('title').value).toEqual('2025-05-05');
      expect(changeListener).not.toHaveBeenCalled();

      const dateInput = document.getElementById('event-date');

      dateInput.dispatchEvent(new Event('cut'));

      jest.runAllTimers();

      expect(document.getElementById('title').value).toEqual('');
      expect(changeListener).toHaveBeenCalledTimes(1);
    });

    it.todo('should allow for no change events to be dispatched', () => {
      /**
       * Set the quiet value and do an apply and clear action to confirm change events are not dispatched
       */
    });
  });

  describe('delayed sync between two fields', () => {
    beforeAll(() => {
      application?.stop();

      document.body.innerHTML = `
    <section>
      <input type="text" name="title" id="title" />
      <input
        type="date"
        id="event-date"
        name="event-date"
        value="2025-07-22"
        data-controller="w-sync"
        data-action="change->w-sync#apply cut->w-sync#clear"
        data-w-sync-target-value="#title"
        data-w-sync-delay-value="500"
      />
    </section>`;

      application = Application.start();
    });

    it.todo(
      'should delay the update on change based on the set value',
      () => {},
    );

    it.todo(
      'should delay the update on apply based on the set value',
      () => {},
    );
  });

  describe('ability for the sync to be disabled between two fields', () => {
    beforeAll(() => {
      application?.stop();

      document.body.innerHTML = `
    <section>
      <input type="text" name="title" id="title" value="keep me"/>
      <input
        type="date"
        id="event-date"
        name="event-date"
        value="2025-07-22"
        data-controller="w-sync"
        data-action="change->w-sync#apply cut->w-sync#clear"
        data-w-sync-target-value="#title"
      />
    </section>`;
    });

    it.todo(
      'should allow for the target element to block syncing at the start',
      () => {
        /**
         * add an event listener to #title that prevents default on `w-sync:start`
         * THEN start the application
         * THEN check that the data attribute `data-w-sync-disabled-value` is true
         * Then try to do a 'apply' to confirm it is not updating
         */

        application = Application.start();
      },
    );

    it.todo(
      'should allow for the target element to block syncing with the check approach',
      () => {
        /**
         * Change the data attribute on the HTML (remember changing attributes is the same as changing state)
         * so that the disabled-value is no longer present.
         * Now do a similar test to above with the 'check' event.
         * Then try to do a 'clear' to confirm it is not updating
         */
      },
    );
  });
});
```